vnext/Microsoft.ReactNative/JsiApi.cpp

```diff
@@ -58,137 +43,212 @@ struct JsiBufferWrapper : facebook::jsi::Buffer {
-  return *reinterpret_cast<facebook::jsi::String const *>(ptr);
-}
+  static JsiSymbolRef MakeJsiSymbolData(facebook::jsi::Symbol &&symbol) {
+    auto &&accessor = AsPointerAccessor(std::move(symbol));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
 
-static facebook::jsi::Object const &AsObject(facebook::jsi::Runtime::PointerValue const **ptr) noexcept {
-  return *reinterpret_cast<facebook::jsi::Object const *>(ptr);
-}
+  static JsiStringRef MakeJsiStringData(facebook::jsi::String &&str) {
+    auto &&accessor = AsPointerAccessor(std::move(str));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
 
-static facebook::jsi::Function const &AsFunction(facebook::jsi::Runtime::PointerValue const **ptr) noexcept {
-  return *reinterpret_cast<facebook::jsi::Function const *>(ptr);
-}
+  static JsiObjectRef MakeJsiObjectData(facebook::jsi::Object &&obj) {
+    auto &&accessor = AsPointerAccessor(std::move(obj));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
 
-static facebook::jsi::WeakObject &AsWeakObject(facebook::jsi::Runtime::PointerValue **ptr) noexcept {
-  return *reinterpret_cast<facebook::jsi::WeakObject *>(ptr);
-}
+  static JsiPropertyIdRef MakeJsiPropertyNameIdData(facebook::jsi::PropNameID &&propertyNameId) {
+    auto &&accessor = AsPointerAccessor(std::move(propertyNameId));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
 
-static facebook::jsi::Array const &AsArray(facebook::jsi::Runtime::PointerValue const **ptr) noexcept {
-  return *reinterpret_cast<facebook::jsi::Array const *>(ptr);
-}
+  static JsiObjectRef MakeJsiArrayData(facebook::jsi::Array &&arr) {
+    auto &&accessor = AsPointerAccessor(std::move(arr));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
 
-static facebook::jsi::ArrayBuffer const &AsArrayBuffer(facebook::jsi::Runtime::PointerValue const **ptr) noexcept {
-  return *reinterpret_cast<facebook::jsi::ArrayBuffer const *>(ptr);
-}
+  static JsiWeakObjectRef MakeJsiWeakObjectData(facebook::jsi::WeakObject &&weakObj) {
+    auto &&accessor = AsPointerAccessor(std::move(weakObj));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
 
-static facebook::jsi::Value const *AsValue(JsiValueRef const &data) noexcept {
-  return reinterpret_cast<facebook::jsi::Value const *>(&data);
-}
+  static JsiObjectRef MakeJsiFunctionData(facebook::jsi::Function &&func) {
+    auto &&accessor = AsPointerAccessor(std::move(func));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
+};
 
-static JsiValueRef ToJsiValueData(facebook::jsi::Value const &value) noexcept {
-  return {static_cast<JsiValueKind>(static_cast<int32_t>(value.kind_)),
-          *reinterpret_cast<uint64_t const *>(&value.data_.number)};
-}
+struct ValueAccessor {
```

Comment:
final

Reply:
I wonder why do we need the `final` keyword for the non-virtual classes? Could you refer me to any guidelines where it is considered as a good practice. IMHO, it is useless and just adds an unnecessary noise. The `final` keyword must not be applied to all classes blindly. It is a very specific tool to help compiler optimizer with de-virtualization.